lib/pharos/ssh/manager.rb

```diff
@@ -0,0 +1,43 @@
+
+module Pharos
+  module SSH
+    class Manager
+      def initialize
+        @clients = {}
+      end
+
+      # @param host [Pharos::Configuration::Host]
+      def client_for(host)
+        @clients[host] ||= Pharos::SSH::Client.new(host.address, host.user, keys: [host.ssh_key_path]).tap do |client|
+          client.connect
+        end
+      end
+
+      # @param hosts [Array<Pharos::Configuration::Host>]
+      # @return [Array<...>]
+      def with_hosts(hosts)
+        threads = hosts.map do |host|
+          # from the main thread
+          client = client_for(host)
+
+          Thread.new do
```

Comment:
Probably would make sense to use thread pool here.

Reply:
Using a thread pool for minimizing `Thread.new` overhead is probably not that relevant here, because phases are pretty heavyweight (remote SSH/API calls).

When running with very large numbers of nodes, you might also want to start limiting the numbers of nodes handled in parallel at some point... but that can also happen later as an optimization, without needing to chance the interface used by `Pharos::UpCommand`.

OTOH the error handling could be a bit more fine-tuned, right now it joins the threads in order, so if some nodes fail with an exception, then it's a bit random as to which error this raises, and how long it waits for the earlier nodes to complete before raising.